bluepysnap/query.py

```diff
@@ -0,0 +1,159 @@
+"""Module to process search queries of nodes/edges."""
+from collections.abc import Mapping
+from copy import deepcopy
+
+import numpy as np
+import pandas as pd
+from bluepysnap.exceptions import BluepySnapError
+from bluepysnap import utils
+
+# this constant is not part of the sonata standard
+NODE_ID_KEY = "node_id"
+EDGE_ID_KEY = "edge_id"
+POPULATION_KEY = "population"
+OR_KEY = "$or"
+AND_KEY = "$and"
+REGEX_KEY = "$regex"
+NODE_SET_KEY = "$node_set"
+VALUE_KEYS = {REGEX_KEY}
+ALL_KEYS = {NODE_ID_KEY, EDGE_ID_KEY, POPULATION_KEY, OR_KEY, AND_KEY, NODE_SET_KEY} | VALUE_KEYS
+
+
+# TODO: move to `libsonata` library
+def _complex_query(prop, query):
+    result = np.full(len(prop), True)
+    for key, value in query.items():
+        if key == '$regex':
+            result = np.logical_and(result, prop.str.match(value + "\\Z"))
+        else:
+            raise BluepySnapError("Unknown query modifier: '%s'" % key)
+    return result
+
+
+def _positional_mask(data, ids):
+    """Positional mask for the node IDs.
+
+    Args:
+        ids (None/numpy.ndarray): the ids array. If None all ids are selected.
+
+    Examples:
+        if the data set contains 5 nodes:
+        _positional_mask([0,2]) --> [True, False, True, False, False]
+    """
+    if ids is None:
+        return np.full(len(data), fill_value=True)
+    mask = np.full(len(data), fill_value=False)
+    valid_ids = pd.Index(utils.ensure_list(ids)).intersection(data.index)
+    mask[valid_ids] = True
+    return mask
+
+
+def _circuit_mask(data, population_name, queries):
+    """Handle the population, node ID queries."""
+    populations = queries.pop(POPULATION_KEY, None)
+    if populations is not None and population_name not in set(utils.ensure_list(populations)):
+        ids = []
+    else:
+        ids = queries.pop(NODE_ID_KEY, queries.pop(EDGE_ID_KEY, None))
+    return queries, _positional_mask(data, ids)
+
+
+def _properties_mask(data, population_name, queries):
+    """Return mask of IDs matching `props` dict."""
+    unknown_props = set(queries) - set(data.columns) - ALL_KEYS
+    if unknown_props:
+        return np.full(len(data), fill_value=False)
+
+    queries, mask = _circuit_mask(data, population_name, queries)
+    if not mask.any():
+        # Avoid fail and/or processing time if wrong population or no nodes
+        return mask
+
+    for prop, values in queries.items():
+        prop = data[prop]
+        if np.issubdtype(prop.dtype.type, np.floating):
+            v1, v2 = values
+            prop_mask = np.logical_and(prop >= v1, prop <= v2)
+        elif isinstance(values, str) and values.startswith('regex:'):
+            prop_mask = _complex_query(prop, {'$regex': values[6:]})
```

Comment:
I think we should remove this part. This is redundant with the Mapping part and I removed it from bleupy also.